src/main/java/de/tum/in/www1/artemis/service/LearningGoalService.java

```diff
@@ -0,0 +1,115 @@
+
+import org.springframework.stereotype.Service;
+
+import de.tum.in.www1.artemis.domain.Exercise;
+import de.tum.in.www1.artemis.domain.LearningGoal;
+import de.tum.in.www1.artemis.domain.Result;
+import de.tum.in.www1.artemis.domain.User;
+import de.tum.in.www1.artemis.domain.lecture.ExerciseUnit;
+import de.tum.in.www1.artemis.domain.lecture.LectureUnit;
+import de.tum.in.www1.artemis.domain.participation.StudentParticipation;
+import de.tum.in.www1.artemis.repository.LearningGoalRepository;
+import de.tum.in.www1.artemis.repository.ResultRepository;
+import de.tum.in.www1.artemis.web.rest.dto.LearningGoalProgress;
+
+@Service
+public class LearningGoalService {
+
+    private final LearningGoalRepository learningGoalRepository;
+
+    private final AuthorizationCheckService authorizationCheckService;
+
+    private final ResultRepository resultRepository;
+
+    private final ParticipationService participationService;
+
+    public LearningGoalService(LearningGoalRepository learningGoalRepository, AuthorizationCheckService authorizationCheckService, ParticipationService participationService,
+            ResultRepository resultRepository) {
+        this.learningGoalRepository = learningGoalRepository;
+        this.authorizationCheckService = authorizationCheckService;
+        this.participationService = participationService;
+        this.resultRepository = resultRepository;
+    }
+
+    /**
+     * Calculate the progress in a learning goal for a specific user
+     * @param learningGoal learning goal to get the progress for
+     * @param user user to get the progress for
+     * @return progress of the user in the learning goal
+     */
+    public LearningGoalProgress getLearningGoalPerformance(LearningGoal learningGoal, User user) {
+
+        LearningGoalProgress learningGoalProgress = new LearningGoalProgress();
+        learningGoalProgress.learningGoalId = learningGoal.getId();
+        learningGoalProgress.learningGoalTitle = learningGoal.getTitle();
+        learningGoalProgress.totalPointsAchievableByStudentsInLearningGoal = 0.0;
+        learningGoalProgress.pointsAchievedByStudentInLearningGoal = 0.0;
+
+        // The progress will be calculated from a subset of the connected lecture units
+        Set<LectureUnit> filteredLectureUnits = learningGoal.getLectureUnits().parallelStream().filter(LectureUnit::isVisibleToStudents).filter(lectureUnit -> {
+            if (lectureUnit instanceof ExerciseUnit) {
+                Exercise exercise = ((ExerciseUnit) lectureUnit).getExercise();
+                return exercise.isAssessmentDueDateOver();
+            }
+            else {
+                return true;
+            }
+        }).collect(Collectors.toSet());
+
+        // In the case that two or more connected exercise units reference the same exercise, only one of them should be used in the progress calculation
+        Set<Exercise> exercisesAlreadyUsedInCalculation = new HashSet<>();
+
```

Comment:
split this into 2 methods

Reply:
Done in da86f2c21800028e9a9a960cd727cf04ede905c7